src/Core/DevOps/StaticAnalyze/PHPStan/Rules/PublicServiceDecoratorRule.php

```diff
@@ -0,0 +1,109 @@
+<?php declare(strict_types=1);
+
+namespace Shopware\Core\DevOps\StaticAnalyze\PHPStan\Rules;
+
+use PhpParser\Node;
+use PhpParser\Node\Stmt\Class_;
+use PHPStan\Analyser\Scope;
+use PHPStan\Rules\Rule;
+use PHPStan\Rules\RuleErrorBuilder;
+use PHPStan\Symfony\ServiceMap;
+use Shopware\Core\Framework\Log\Package;
+
+/**
+ * @internal
+ *
+ * @implements Rule<Class_>
+ */
+#[Package('framework')]
+class PublicServiceDecoratorRule implements Rule
+{
+    public function __construct(private readonly ServiceMap $serviceMap)
+    {
+    }
+
+    public function getNodeType(): string
+    {
+        return Class_::class;
+    }
+
+    public function processNode(Node $node, Scope $scope): array
+    {
+        if (!$node instanceof Class_) {
+            return [];
+        }
+
+        $className = $node->namespacedName?->name;
+        if (!$className) {
+            return [];
+        }
+
+        $service = $this->serviceMap->getService($className);
```

Comment:
getting the service should be after the test class check to avoid unnecessary operation